### DISCUSSION TITLE
Qualificaiton Roadmap

### DISCUSSION THREAD

**Discussion by kodesiba** [Gilead-Public/gsm#801] [General]
# `Make_Snapshot()` Qualification 

In an effort to drive discussion on how to qualify `Make_Snapshot()` in a way that is complete but not overly complicated I have broken the elements of the function and the sub functions it calls into the following categories.  Each of these had a group of functions under them that do different things for each assessment or workflow that is being created or called.  Let me know if anyone has other thoughts or if we need to structure out more pieces or even build out a use cases document for the levels of abstraction to make sure we are covering everything.

## Metadata

While metadata is a big part of `Make_Snapshot()` there's not a ton of functions for it.  The main thing it does is read in metadata files, these could also be used on the qualification side if that is the most complete way to qualify this piece.  In addition to overall metadata, there are additional metadata checking functions used at the `Assess_` level that may need to be added to the qualification.  Every function in the `Assess_` family has a `lChecks` output object that includes the metadata passed in and the output of the checks done on it. Most of this looks to be driven off `is_mapping_valid()` so that would be the route for qualification if needed

## Map/Assess/Analyze/Visualize

These functions are the core of what is going on and are what compute and display the actual statistical elements.  Currently, the `Assess_` and `Analyze_` families of functions are qualified.  The addition of new statistical methods and bounds predictions will be qualified as they are added.  While the `Map_` family is not qualified yet, I believe that this will be straightforward and is something worth qualifying.  The final piece is the `Visualize_` family of functions, these plot the results of the statistical analysis done.  If these visualizations will be used as direct outputs in the user-facing application they would definitely be worth qualifying.

## Utility

Throughout `Make_Snapshot()` there are many utility functions that are called such as wrappers or mapping functions for building or executing workflows.  These functions include utilities such as `MakeAssessmentList()` and `RunAssessment()` which are there to execute other functions more easily.  Additionally things like `Study_Assess()` are run to run assessments on a study level.  I think that while working out from the `Map_` and `Assess_` functions there could be an easy place too qualify these.

## Roadmap
My proposed way forward for qualification is as follows and I think aligns with what we are trying to achieve.  This doesn't currently include the updates to stat methods or bounds prediction that will be added as they are finalized.

1. Create specs and test cases to qualify `Map_` family of functions.  I think that for this we could either use the mapping_rawplus.yaml file to drive it or qualification could create independent mapping.  I don't think either way is easier it's just what we want to qualify and the best way to do this
2. If stats functions are updated, add new stat methods and bounds prediction to fully qualify the `lData` output object.  If not begin next step and revisit when ready.
3. Begin abstraction of qualification through utility functions starting with `RunAssessment()` and consequently `RunStep()`.  
4. Abstract to the combining of multiple assessments into a workflow through qualifying `Study_Assess()` and possibly `MakeAssessmentList()` if the metadata for these workflows needs to be qualified. 
5. Finally, qualify `Make_Snapshot()` through putting all these pieces together along with either created metadata for qualification or using the metadata already specified

**Comment by jwildfire** [comment]
I think this strategy sounds good! Definitely makes sense to prioritize the map functions. The utility/meta functions are mostly workflow oriented. We can work on creating a short list of data manipulations included in those functions, and validate those, and then confirm that no other data is changed.

**Comment by collleenmclaughlin** [comment]
@chelseadickens @collleenmclaughlin
